crates/lang_srv/src/server.rs

```diff
@@ -187,22 +301,189 @@ impl LanguageServer for RocLs {
             partial_result_params: _,
         } = params;
 
-        panic_wrapper(|| self.registry().semantic_tokens(&text_document.uri))
+        panic_wrapper_async(|| async {
+            self.inner
+                .registry()
+                .await
+                .semantic_tokens(&text_document.uri)
+                .await
+        })
+        .await
+    }
+    async fn completion(&self, params: CompletionParams) -> Result<Option<CompletionResponse>> {
+        let doc = params.text_document_position;
+        trace!("got completion request");
+        let res = panic_wrapper_async(|| async {
+            self.inner
+                .registry
+                .completion_items(&doc.text_document.uri, doc.position)
+                .await
+        })
+        .await;
+        res
     }
 }
 
-fn panic_wrapper<T>(f: impl FnOnce() -> Option<T> + std::panic::UnwindSafe) -> Result<Option<T>> {
+async fn panic_wrapper_async<Fut, T>(
+    f: impl FnOnce() -> Fut + std::panic::UnwindSafe,
+) -> Result<Option<T>>
+where
+    Fut: Future<Output = Option<T>>,
+{
     match std::panic::catch_unwind(f) {
-        Ok(r) => Ok(r),
+        Ok(r) => Ok(r.await),
         Err(_) => Err(tower_lsp::jsonrpc::Error::internal_error()),
     }
 }
 
-#[tokio::main]
+#[tokio::main(flavor = "multi_thread")]
```

Comment:
Nit: Maybe let's cap at 2 threads

Reply:
This was actually pointless because it defaults to multi_thread anyway. I think it's a good idea to let it run with lots of threads for the sake of allowing many analysis threads